source/MaterialXShaderGen/ShaderGenerators/Glsl/GlslShaderGenerator.cpp

```diff
@@ -339,10 +363,45 @@ void GlslShaderGenerator::emitFunctionDefinitions(Shader& shader)
     _bsdfNodeArguments[0] = Argument("vec3", "wi");
     _bsdfNodeArguments[1] = Argument("vec3", "wo");
 
-    // Emit function for handling texture coords v-flip 
-    // as needed by the v-direction set by the user
     BEGIN_SHADER_STAGE(shader, HwShader::PIXEL_STAGE)
+
+        // Emit function for handling texture coords v-flip 
+        // as needed by the v-direction set by the user
         shader.addBlock(shader.getRequestedVDirection() != getTargetVDirection() ? VDIRECTION_FLIP : VDIRECTION_NOOP);
+
+        // For surface shaders we need light shaders
+        if (shader.hasClassification(SgNode::Classification::SHADER | SgNode::Classification::SURFACE))
+        {
```

Comment:
This is where light handling is emitted. One function for each bound light shader, one function for number of active light sources, and one function for picking the right shader based on light type.